deploy.py

```diff
@@ -284,84 +468,227 @@ def verify_flashed_app(self, expected_app):
+      # Process kernel
+      kernel_path = os.path.join(board_props.path, "target", board_props.arch,
+                                 "release", "{}.bin".format(self.args.board))
+      with open(kernel_path, "rb") as kernel:
+        kern_hex = intelhex.IntelHex()
+        kern_hex.frombytes(kernel.read(), offset=board_props.kernel_address)
+        final_hex.merge(kern_hex, overlap="error")
+
+    if self.args.application:
+      # Add padding
+      if board_props.padding_address:
+        padding_hex = intelhex.IntelHex()
+        padding_hex.frombytes(
+            self.get_padding(), offset=board_props.padding_address)
+        final_hex.merge(padding_hex, overlap="error")
+
+      # Now we can add the application from the TAB file
+      app_tab_path = "target/tab/{}.tab".format(self.args.application)
+      assert os.path.exists(app_tab_path)
+      app_tab = tab.TAB(app_tab_path)
+      if board_props.arch not in app_tab.get_supported_architectures():
+        fatal(("It seems that the TAB file was not produced for the "
+               "architecture {}".format(board_props.arch)))
+      app_hex = intelhex.IntelHex()
+      app_hex.frombytes(
+          app_tab.extract_app(board_props.arch).get_binary(),
+          offset=board_props.app_address)
+      final_hex.merge(app_hex)
+    info("Generating all-merged HEX file: {}".format(dest_file))
+    final_hex.tofile(dest_file, format="hex")
+
+  def check_prerequisites(self):
+    if self.args.programmer == "jlink":
+      assert_mandatory_binary("JLinkExe")
+
+    if self.args.programmer == "openocd":
+      assert_mandatory_binary("openocd")
+
+    if self.args.programmer == "pyocd":
+      assert_mandatory_binary("pyocd")
+      assert_python_library("intelhex")
+      if not SUPPORTED_BOARDS[self.args.board].pyocd_target:
+        fatal("This board doesn't seem to support flashing through pyocd.")
+
+    if self.args.programmer == "nordicdfu":
+      assert_mandatory_binary("nrfutil")
+      assert_python_library("intelhex")
+      assert_python_library("nordicsemi.lister")
+      if not SUPPORTED_BOARDS[self.args.board].nordic_dfu:
+        fatal("This board doesn't support flashing over DFU.")
+
+    if self.args.programmer == "none":
+      assert_python_library("intelhex")
+
   def run(self):
-    if self.args.action is None:
-      # Nothing to do
+    if self.args.listing == "boards":
+      print(os.linesep.join(get_supported_boards()))
       return 0
 
+    if self.args.listing == "programmers":
+      print(os.linesep.join(PROGRAMMERS))
+      return 0
+
+    if self.args.listing:
+      # Missing check?
+      fatal("Listing {} is not implemented.".format(self.args.listing))
+
+    self.check_prerequisites()
     self.update_rustc_if_needed()
 
-    if self.args.action == "os":
-      info("Installing Tock on board {}".format(self.args.board))
-      self.build_and_install_tockos()
+    if not self.args.tockos and not self.args.application:
+      info("Nothing to do.")
```

Comment:
I'd return a non-zero error code in that case so that people are aware they need to provide something to do in their command line?